Ocamldep: -shared option

Octachron opened the pull request:
This short PR introduces a new ocamldep option: `-shared`. When this option is enabled, ocamldep generates dependencies for plugin files in complement of the classical `.cmx` targets (if native dependencies are enabled).

The second commit of this PR uses this new option to update ocamldoc makefile (see #1031).

Octachron committed:
-shared option for ocamldep

shindere commented:
The PR looks good to me, excepts it is missing a Changes entry and I
guess the ocamldep man page would need to be updated, as perhaps the
OCaml manual.

Octachron committed:
Document ocamldep's "-shared" option

Octachron committed:
GPR#1045 changes

Octachron commented:
I have updated the PR to add the documentation and a change entry (in the 4.06 section for now?).

xavierleroy commented:
Looks good and low-risk to me.  @damiendoligez will tell us if he wants it in 4.05.  

Concerning the documentation, maybe "in addition to" would be better than "on top of", especially if you mean "en plus de".  @dra27 or @mshinwell can suggest appropriate British wordings.

gasche commented:
I think that we don't need to bother putting this in 4.05 (it's not very a high-priority feature), but we should definitely cherry-pick d620cd486e69064c39fcb0c0b839ae827d134b27 in all open branches to fix the build problems.

gasche commented:
Hm, in fact d620cd486e69064c39fcb0c0b839ae827d134b27 contains both the change to `depend:` rule in the `Makefile`, and the actual changes in `.depend`. So we either have to split the two to cherry-pick the `.depend` changes only, or cherry-pick the whole branch if we want `make depend` to work correctly.

dra27 commented:
I agree with @xavierleroy, "in addition to" would be better. That said, wouldn't it be even better to have `-shared` imply `-native`?

dra27 commented:
Actually, ignore that last bit - `-shared` of course doesn't imply `native`, it just conflicts with `-bytecode`, but ocamldep already doesn't complain about that elsewhere.

Octachron committed:
Update ocamldoc Makefile

Octachron committed:
Update ocamldoc .depend

Octachron commented:
"In addition to" definitely sounds better than "on top of", I have updated the documentation. Similarly, I have split in two the update to ocamldoc 's Makefile and .depend file, for an easier cherry-picking of the latter change.

dra27 commented:
All looks good to me - @gasche, https://github.com/ocaml/ocaml/pull/1045/commits/f6736c18ceaf091cd95e2467fa4965fe983b6ec8 should be cherry-pick-able now?

xavierleroy commented:
It goes on trunk right now.

xavierleroy commented:
Cherry-picked [ f6736c1 ] in 4.05 and 4.04.  I didn't test at all...

dra27 commented:
CI confirms success on all branches! ([AppVeyor](https://ci.appveyor.com/project/avsm/ocaml/history) & [Travis](https://travis-ci.org/ocaml/ocaml/builds))
